samplemv/samplemv.go

```diff
@@ -0,0 +1,281 @@
+	Q   distmv.Quantiler
+	Src *rand.Rand
+}
+
+// Sample generates rows(batch) samples using the LatinHypercube generation
+// procedure.
+func (l LatinHypercuber) Sample(batch *mat64.Dense) {
+	LatinHypercube(batch, l.Q, l.Src)
+}
+
+// LatinHypercube generates rows(batch) samples using Latin hypercube sampling
+// from the given distribution. If src != nil, it will be used to generate
+// random numbers, otherwise rand.Float64 will be used.
+//
+// Latin hypercube sampling divides the cumulative distribution function into equally
+// spaced bins and guarantees that one sample is generated per bin. Within each bin,
+// the location is randomly sampled. The distmv.UnitNormal variable can be used
+// for easy generation from the unit interval.
+func LatinHypercube(batch *mat64.Dense, q distmv.Quantiler, src *rand.Rand) {
+	r, c := batch.Dims()
+	var f64 func() float64
+	var perm func(int) []int
+	if src != nil {
+		f64 = src.Float64
+		perm = src.Perm
+	} else {
+		f64 = rand.Float64
+		perm = rand.Perm
+	}
+	r64 := float64(r)
+	for i := 0; i < c; i++ {
+		p := perm(r)
+		for j := 0; j < r; j++ {
+			var v float64
+			v = f64()
+			v = v/r64 + float64(j)/r64
+			batch.Set(p[j], i, v)
+		}
+	}
+	p := make([]float64, c)
+	for i := 0; i < r; i++ {
+		copy(p, batch.RawRowView(i))
+		q.Quantile(batch.RawRowView(i), p)
+	}
+}
+
+// Importancer is a wrapper around the Importance sampling generation method.
+type Importancer struct {
+	Target   distmv.LogProber
+	Proposal distmv.RandLogProber
+}
+
+// Sample generates rows(batch) samples using the Importance sampling generation
+// procedure.
+func (l Importancer) SampleWeighted(batch *mat64.Dense, weights []float64) {
+	Importance(batch, weights, l.Target, l.Proposal)
+}
+
+// Importance sampling generates rows(batch) samples from the proposal distribution,
+// and stores the locations and importance sampling weights in place.
+//
+// Importance sampling is a variance reduction technique where samples are
+// generated from a proposal distribution, q(x), instead of the target distribution
+// p(x). This allows relatively unlikely samples in p(x) to be generated more frequently
+//
+// The importance sampling weight at x is given by p(x)/q(x). To reduce variance,
+// a good proposal distribution will bound this sampling weight. This implies the
+// support of q(x) should be at least as broad as p(x), and q(x) should be "fatter tailed"
+// than p(x).
+//
+// If weights is nil the weights are not stored. The length of weights must equal
+// the length of batch, otherwise Importance will panic.
+func Importance(batch *mat64.Dense, weights []float64, target distmv.LogProber, proposal distmv.RandLogProber) {
+	r, _ := batch.Dims()
+	if r != len(weights) {
+		panic(badLengthMismatch)
+	}
+	for i := 0; i < r; i++ {
+		v := batch.RawRowView(i)
+		proposal.Rand(v)
+		weights[i] = math.Exp(target.LogProb(v) - proposal.LogProb(v))
+	}
+}
+
+var ErrRejection = errors.New("rejection: acceptance ratio above 1")
```

Comment:
Add a comment to make golint happy? Also, should the string start with samplemv:, something like "samplemv: acceptance ratio above 1 in Rejection"?